### PR TITLE
docs: add page on hyper parameters

### PR DESCRIPTION
**Pull request by LMMilliken**
This pr adds content to the hyper parameters page of the new documentation structure, explaining hyper parameters such as learning rate, epochs and batch size, as well as pointing users to other pages in the documentation for more complex parameters such as optimizers and loss functions

---

- [ ] This PR references an open issue
- [ ] I have added a line about this change to CHANGELOG

### CODE REVIEW

File: docs/concepts/hyperparameters.md [hunk 1]
```diff
@@ -1,5 +1,103 @@
...
+
+run = finetuner.fit(
+    model='efficientnet_b0',
+    train_data=train_data,
+    eval_data=eval_data, 
+    run_name='finetune-flickr-dataset-efficientnet-1',
+    description='this is a trial run on flickr8k dataset with efficientnet b0.',
+    experiment_name='finetune-flickr-dataset', # Link to the experiment created above.
+    model_options={}, # Additional options to pass to the model constructor.
+    loss='TripletMarginLoss', # Use CLIPLoss for CLIP fine-tuning.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:26]:
```suggestion
    loss='TripletMarginLoss',
```

File: docs/concepts/hyperparameters.md [hunk 2]
```diff
@@ -1,5 +1,103 @@
...
+    miner='TripletMarginMiner',
+    miner_options={'margin': 0.2}, # Additional options for the miner constructor.
+    scheduler='linear', # Use a linear scheduler to adjust the learning rate.
+    scheduler_options={}, # Additional options for the scheduler.
+    optimizer='Adam',
+    optimizer_options={'weight_decay': 0.01}, # Additional options for the optimizer.
+    learning_rate = 1e-4,
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:33]:
```suggestion
    learning_rate=1e-4,
```

File: docs/concepts/hyperparameters.md [hunk 3]
```diff
@@ -1,5 +1,103 @@
...
+    epochs=5,
+    batch_size=128,
+    scheduler_step='batch',
+    freeze=False, # If applied will freeze the embedding model, only train the MLP.
+    output_dim=512, # Attach a MLP on top of embedding model.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:38]:
```suggestion
    output_dim=512, # Attach an MLP on top of embedding model.
```

File: docs/concepts/hyperparameters.md [hunk 4]
```diff
@@ -1,5 +1,103 @@
...
+    device='cuda',
+    num_workers=4,
+    to_onnx=False,  # If set, please pass `is_onnx` when making inference.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:41]:
```suggestion
    to_onnx=False,  # If set, please pass `is_onnx` when doing inference.
```

File: docs/concepts/hyperparameters.md [hunk 5]
```diff
@@ -1,5 +1,103 @@
...
+    csv_options=CSVOptions(),  # Additional options for reading data from a CSV file.
+    public=False,  # If set, anyone has the artifact id can download your fine-tuned model.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:43]:
```suggestion
    public=False,  # If set, anyone who has the artifact id can download your fine-tuned model.
```

File: docs/concepts/hyperparameters.md [hunk 6]
```diff
@@ -1,5 +1,103 @@
...
+    num_items_per_class=4,  # How many items per class to include in a batch.
+)
+```
+
+Several of these parameters are explained in detail in their own sections,
+the rest of this section will explain how best to use the ones that aren't.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:49]:
```suggestion
the rest of this section will explain how to best use the ones that aren't.
```

File: docs/concepts/hyperparameters.md [hunk 7]
```diff
@@ -1,5 +1,103 @@
...
+a higher learning rate results in a greater change.
+It is important to choose an appropriate learning rate when training;
+too high and you will overfit on your training data and you will lose the knowledge gained from pre-training,
+too low and the model will not chage enough to produce significant improvement.  
+Typically, an apropriate learning rate lies in the range of `1e-4` to `1e-6`. To see a typical learning rate for the different models and tasks we support, check out our [examples](#todo: figure out what to put here).
+
+## Epochs
+Each epoch is a single round of training, in each of these epochs, the model is trained on the entire training
+data, so training for a higher number of epochs will result in a larger change in the model's performance.
+Typically 2-3 epochs are enough for a fine-tuning job.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:62]:
```suggestion
Typically 3-10 epochs are enough for a fine-tuning job.
```
more like 3 to 10 based on what we do in the examples

File: docs/concepts/hyperparameters.md [hunk 8]
```diff
@@ -1,5 +1,103 @@
...
+in out of memory errors. Typically, a `batch_size` of 64 or 128 are good options when you
+are unsure of how high you can set this value, however you can also choose to not set the `batch_size`
+at all, in which case the highest possible value will be calculated for you automatically.
+
+## Learning Rate Scheduler
+You can configure Finetuner to use a learning rate scheduler.
+The scheduler is used to adjust the learning rate during training.
+If no scheduler is configured, the learning rate is constant during training.
+When a scheduler is configured, the learning rate is adjusted after each batch by default.
+Alternatively, one can set `scheduler_options = {'scheduler_step': 'epoch'}` to adjust the learning rate after
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:77]:
```suggestion
Alternatively, you can set `scheduler_options = {'scheduler_step': 'epoch'}` to adjust the learning rate after
```

File: docs/concepts/hyperparameters.md [hunk 9]
```diff
@@ -1,5 +1,103 @@
...
+each epoch, when no scheduler is configured, this `scheduler_options` parameter is ignored.
+A scheduler usually has a warm-up phase, where the learning rate is increasing.
+After that most learning rateschedulers decrease the learning rate.
+For example, the `linear` scheduler decreases the learning rate linearly from the initial learning rate:
+![Learning Rate](https://user-images.githubusercontent.com/6599259/221238105-ee294b7e-544a-4de8-8c92-0c61275f29bb.png)  
+The length of the warm-up phase is configured via the `num_warmup_steps` option inside `scheduler_optons`.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:83]:
```suggestion
The length of the warm-up phase is configured via the `num_warmup_steps` option inside `scheduler_options`.
```

File: docs/concepts/hyperparameters.md [hunk 10]
```diff
@@ -1,5 +1,103 @@
...
+The loss function determines the training objective.
+The type of loss function which is most suitable for your task depends heavily on the task your training for.
+The `miner` parameter is used to specify the method used to construct batches.
+Loss functions can have different requirements for the contents of each batch,
+so the available miners can depend on the loss function used.
+
+See the [loss functions](./loss-functions.md) page to see the available loss functions.
+
+## Optimizer
+The `optimizer` is the method used to update the weights of the model after each training batch.
```
Comment (gmastrapas) [inline docs/concepts/hyperparameters.md:100]:
```suggestion
The `optimizer` determines the method used to update the weights of the model after each training batch.
```

**Review by gmastrapas — CHANGES_REQUESTED** [review]

File: docs/concepts/hyperparameters.md [hunk 11]
```diff
@@ -1,5 +1,103 @@
 (hyperparameters)=
 # {octicon}`tools` Hyperparameters
 
-Explain the influence of number of epochs, batch size, for learning rate - reference to optimizer documentation
+Beyond the simplest use case,
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:4]:
```suggestion
If needed,
```

File: docs/concepts/hyperparameters.md (hunk 6 above)
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:49]:
```suggestion
the rest of this section will explain how to use the ones that aren't.
```

File: docs/concepts/hyperparameters.md [hunk 12]
```diff
@@ -1,5 +1,103 @@
...
+    num_items_per_class=4,  # How many items per class to include in a batch.
+)
+```
+
+Several of these parameters are explained in detail in their own sections,
+the rest of this section will explain how best to use the ones that aren't.
+
+## Learning Rate
+The learning rate affects how much new training data affects the weights of the model you are training,
+a higher learning rate results in a greater change.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:53]:
```suggestion
The learning rate affects how much each item of new training data can affect the weights of the model you are training each time the model sees it.
A higher learning rate means faster changes to the model, with greater change to the weights for each training epoch. A lower rate means changes will be slower.
```

File: docs/concepts/hyperparameters.md [hunk 13]
```diff
@@ -1,5 +1,103 @@
...
+It is important to choose an appropriate learning rate when training;
+too high and you will overfit on your training data and you will lose the knowledge gained from pre-training,
+too low and the model will not chage enough to produce significant improvement.  
+Typically, an apropriate learning rate lies in the range of `1e-4` to `1e-6`. To see a typical learning rate for the different models and tasks we support, check out our [examples](#todo: figure out what to put here).
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:57]:
```suggestion
It is essential to choose an appropriate learning rate when training:
too high, and you risk unstable model weights that never converge, overfitting your training data, and you may lose the knowledge gained from pre-training,
too low, and the model will take too long to learn and may not change enough to produce significant improvement.  
An appropriate learning rate is typically somewhere between `1e-4` and `1e-6`. To see typical learning rates for various models and tasks, check out our [examples](#todo: figure out what to put here).
```

File: docs/concepts/hyperparameters.md [hunk 14]
```diff
@@ -1,5 +1,103 @@
...
+
+## Epochs
+Each epoch is a single round of training, in each of these epochs, the model is trained on the entire training
+data, so training for a higher number of epochs will result in a larger change in the model's performance.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:61]:
```suggestion
An epoch is a single round of training in which the model is presented with each item of training data, and after each item, weights are updated. Training for more epochs will result in a larger change in the model's performance but will take more time.
```

File: docs/concepts/hyperparameters.md (hunk 13 above)
Comment (samsja) [inline docs/concepts/hyperparameters.md:57]: I personally feel that asking users to pick a good LR is too much. Like how could they know? I would expect to have an auto lr finder or do an ensemble of multiple runs with different lr.

 But this is outside of the scope of Documentation I guess

File: docs/concepts/hyperparameters.md [hunk 15]
```diff
@@ -1,5 +1,103 @@
...
+During fine-tuning, your training data is split up into batches and is trained on one batch at a time.
+The `batch_size` parameter can be used to configure the size of each batch.
+A larger `batch_size` results in faster training, though too large a `batch_size` can result
+in out of memory errors. Typically, a `batch_size` of 64 or 128 are good options when you
+are unsure of how high you can set this value, however you can also choose to not set the `batch_size`
+at all, in which case the highest possible value will be calculated for you automatically.
+
+## Learning Rate Scheduler
+You can configure Finetuner to use a learning rate scheduler.
+The scheduler is used to adjust the learning rate during training.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:74]:
```suggestion
A learning rate scheduler adjusts the learning rate over the course of training.
This can speed up training under some conditions.
You can configure Finetuner to use a learning rate scheduler by setting the `scheduler` parameter.
```

This would also be an excellent place to say what schedulers are available!

File: docs/concepts/hyperparameters.md [hunk 16]
```diff
@@ -1,5 +1,103 @@
...
+If no scheduler is configured, the learning rate is constant during training.
+When a scheduler is configured, the learning rate is adjusted after each batch by default.
+Alternatively, one can set `scheduler_options = {'scheduler_step': 'epoch'}` to adjust the learning rate after
+each epoch, when no scheduler is configured, this `scheduler_options` parameter is ignored.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:78]:
```suggestion
each epoch. If no scheduler has been configured, the `scheduler_options` parameter is ignored.
```

File: docs/concepts/hyperparameters.md [hunk 17]
```diff
@@ -1,5 +1,103 @@
...
+![Learning Rate](https://user-images.githubusercontent.com/6599259/221238105-ee294b7e-544a-4de8-8c92-0c61275f29bb.png)  
+The length of the warm-up phase is configured via the `num_warmup_steps` option inside `scheduler_optons`.
+By default, it is set to zero.
+
+```{Important}
+Please check the [developer reference](../../api/finetuner/#finetuner.fit) to get the available options for `scheduler`.
+```
+
+## Loss Functions and Miners
+The loss function determines the training objective.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:90]:
```suggestion
A loss function measures how far a model's output is from the correct values.
```

File: docs/concepts/hyperparameters.md [hunk 18]
```diff
@@ -1,5 +1,103 @@
...
+The type of loss function which is most suitable for your task depends heavily on the task your training for.
+The `miner` parameter is used to specify the method used to construct batches.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:92]: I don't understand the role of the miner here, or what it has to do with loss functions.

File: docs/concepts/hyperparameters.md [hunk 19]
```diff
@@ -1,5 +1,103 @@
...
+Loss functions can have different requirements for the contents of each batch,
+so the available miners can depend on the loss function used.
+
+See the [loss functions](./loss-functions.md) page to see the available loss functions.
+
+## Optimizer
+The `optimizer` is the method used to update the weights of the model after each training batch.
+The `optimizer` parameter, in combination with its accompanying `optimizer_options`
+parameter can be configured in a number of ways.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:112]:
```suggestion
The `optimizer` parameter, in combination with the `optimizer_options` parameter, can be configured in several ways.
```

File: docs/concepts/hyperparameters.md (hunk 4 above)
Comment (gmastrapas) [reply docs/concepts/hyperparameters.md]: @LMMilliken

File: docs/concepts/hyperparameters.md [hunk 20]
```diff
@@ -1,5 +1,121 @@
...
+A loss function measures how far a model's output is from the correct values.
+The loss function most suitable for your task depends heavily on the task you're training for.
+
+See the [loss functions](./loss-functions.md) page to see the available loss functions.
+
+## Miner
+Depending on the loss function, the contents of each batch must meet certain requirements,
+for example, the `TripletMarginLoss` function requires at least two of each class to be present in the batch.
+The `miner` is responsible for creating these batches,
+and there are several different miners that each use different techniques to construct these batches.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:110]:
```suggestion
The `miner` is responsible for creating these batches so that they match the loss function's requirements.
```

**Comment by LMMilliken** [comment]
I had a discussion with @guenthermi and he suggested that the code block showing the finetuner.fit function is unnecessary as it repeats what will be in the `The Fit Function` page.

File: docs/concepts/hyperparameters.md [hunk 21]
```diff
@@ -1,5 +1,29 @@
 (hyperparameters)=
 # {octicon}`tools` Hyperparameters
 
-Explain the influence of number of epochs, batch size, for learning rate - reference to optimizer documentation
+Several of the parameters of the `finetuner.fit` function are considered hyper-parameters,
+values that do not change the methods used during training,
+but rather control the overall behavior and settings of the training process.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:6]:
```suggestion
Several of the `finetuner.fit` function's parameters are _hyper-parameters_: values that do not change the methods used during training but control the overall behavior and settings of the training process.
```

File: docs/concepts/hyperparameters.md [hunk 22]
```diff
@@ -1,5 +1,29 @@
...
+and after each item, weights are updated.
+Training for more epochs will result in a larger change in the model's performance but will take more time.
+Training on too many epochs can result in overfitting, so it is important to choose a reasonable number.
+For transformer models, training can be completed with only 1-3 epochs while still producing a good improvement, whereas other models might profit from training on more epochs, typically between 4 and 10.
+
+## Batch Size
+During fine-tuning, training data is split into batches, and training takes place one batch at a time.
+The `batch_size` parameter sets the size of each batch.
+A larger `batch_size` results in faster training, though too large a `batch_size` can result
+in out of memory errors.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:20]:
```suggestion
in out-of-memory errors.
```

**Comment by bwanglzu** [comment]
feel free to go with Scott, George or Michael, they're better reviewers for docs change :)

File: docs/concepts/hyperparameters.md [hunk 23]
```diff
@@ -1,5 +1,32 @@
...
 # {octicon}`tools` Hyperparameters
 
-Explain the influence of number of epochs, batch size, for learning rate - reference to optimizer documentation
+Several of the `finetuner.fit` function's parameters are _hyper-parameters_: values that do not change the methods used during training but control the overall behavior and settings of the training process.
+These fall into four areas:
+- [Setting the learning rate and using an optimizer.](./optimizers.md)
+- [Setting a loss function and using a miner.](./loss-functions.md)
+- Setting the batch size.
+- Choosing how many epochs to train.
+For the first two, see the pages linked above. For epochs and batch sizes, read the discussion below.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:10]:
```suggestion

For the first two, see the pages linked above. For epochs and batch sizes, read the discussion below.
```

I see a formatting problem in the deployed version.

File: docs/concepts/hyperparameters.md [hunk 24]
```diff
@@ -1,5 +1,32 @@
...
+The `batch_size` parameter sets the size of each batch.
+A larger `batch_size` results in faster training, though too large a `batch_size` can result
+in out-of-memory errors.
+The optimal `batch_size` is dependent on the model you are using and the contents of your training data.
+A `batch_size` of 64 or 128 is generally reasonably safe if you don't know how high you can set this value. However, if you do not set the `batch_size` at all,
+Finetuner will determine the highest possible value for your system and set it automatically.
+
+```{Important}
+CLIP models are usually larger than other models, with the `vit-large-en` model only being able to be trained safely with a `batch_size` of 8.  
+If you are not sure what `batch_size` to use for a model, we recommend not setting it and allowing the `batch_size` to be calculated automatically.
```
Comment (scott-martens) [inline docs/concepts/hyperparameters.md:30]:
```suggestion
CLIP models are usually larger than other models, with the `vit-large-en` model only being able to be trained safely with a `batch_size` of 8. If you are not sure what `batch_size` to use for a model, we recommend not setting it and allowing the `batch_size` to be calculated automatically.
```

Same. Formatting problem.

File: docs/concepts/hyperparameters.md [hunk 25]
```diff
@@ -1,5 +1,32 @@
...
+- [Setting the learning rate and using an optimizer.](./optimizers.md)
+- [Setting a loss function and using a miner.](./loss-functions.md)
+- Setting the batch size.
+- Choosing how many epochs to train.
+
+For the first two, see the pages linked above. For epochs and batch sizes, read the discussion below.
+
+## Epochs
+An epoch is a single round of training in which the model is presented with each item of training data,
+and after each item, weights are updated.
```
Comment (guenthermi) [inline docs/concepts/hyperparameters.md:15]:
```suggestion
and during which the model's weights are updated multiple times.
```
the weights are not updated after processing each item, but only after processing a whole batch. If you use gradient accumulation (might be supported in the future) even more rarely.

File: docs/concepts/hyperparameters.md [hunk 26]
```diff
@@ -1,5 +1,32 @@
...
+During fine-tuning, training data is split into batches, and training takes place one batch at a time.
+The `batch_size` parameter sets the size of each batch.
+A larger `batch_size` results in faster training, though too large a `batch_size` can result
+in out-of-memory errors.
+The optimal `batch_size` is dependent on the model you are using and the contents of your training data.
+A `batch_size` of 64 or 128 is generally reasonably safe if you don't know how high you can set this value. However, if you do not set the `batch_size` at all,
+Finetuner will determine the highest possible value for your system and set it automatically.
+
+```{Important}
+CLIP models are usually larger than other models, with the `vit-large-en` model only being able to be trained safely with a `batch_size` of 8. If you are not sure what `batch_size` to use for a model, we recommend not setting it and allowing the `batch_size` to be calculated automatically.
```
Comment (guenthermi) [inline docs/concepts/hyperparameters.md:30]: It is either called `clip-large-en` or `openai/clip-vit-large-patch14` but not vit-large-en

File: docs/concepts/hyperparameters.md [hunk 27]
```diff
@@ -1,5 +1,32 @@
 (hyperparameters)=
 # {octicon}`tools` Hyperparameters
 
-Explain the influence of number of epochs, batch size, for learning rate - reference to optimizer documentation
+Several of the `finetuner.fit` function's parameters are _hyper-parameters_: values that do not change the methods used during training but control the overall behavior and settings of the training process.
```
Comment (guenthermi) [inline docs/concepts/hyperparameters.md:4]: I think it is possible to link the fit function from the API reference

**Review by guenthermi — CHANGES_REQUESTED** [review]
Good work, and thank you for added my missing changes :sweat_smile: . Added some small comments

File: docs/concepts/hyperparameters.md [hunk 28]
```diff
@@ -1,5 +1,35 @@
 (hyperparameters)=
 # {octicon}`tools` Hyperparameters
 
-Explain the influence of number of epochs, batch size, for learning rate - reference to optimizer documentation
+Several of the [`finetuner.fit`](../../api/finetuner/#finetuner.fit) function's parameters are
```
Comment (guenthermi) [inline docs/concepts/hyperparameters.md:4]:
```suggestion
Several of the {meth}`~finetuner.fit` function's parameters are
```

**Review by guenthermi — APPROVED** [review]
one one last comment
